### PR TITLE
[Blazor] Try to load the assembly when finding out a root component or a parameter in webassembly

### LINKED ISSUE #52129: RenderMode not working with razor class libraries
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When I create a project that has a razor class library and I try to use this code
@rendermode InteractiveAuto
and try to compile that library, it always fails to compile because does not know anything about @rendermode

Works fine when using under blazorwebapp.
However, some cases i need razor class library so I can share with blazor hybrid and use other places as well.

### Expected Behavior

The expected behavior is for the project to compile and run.
When using web, then do as promised on the counter where it first uses blazor server and the next time web assembly (since the components would have been downloaded).

### Steps To Reproduce

1.  Create blank solution
2.  Create blazor web app
3. Create maui blazor hybrid
4. Create shared library
5. Move the Counter from blazor web app client to the shared
6. Make the necessary changes necessary including referencing the proper projects and putting in the additional assemblies
7. Run
8. Will always fail to load (even the blazor web app).

### Exceptions (if any)

Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	CS0103	The name 'InteractiveAuto' does not exist in the current context

### .NET Version

8

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by javiercn**
# Load assembly when finding root component or parameter in WebAssembly

This change allows the system to load the assembly when a root component is defined in an RCL and rendered in WebAssembly, even if the assembly is not yet loaded into memory.

## Description

When a root component is defined in an RCL and rendered in WebAssembly, it might happen that the assembly for the component is not yet loaded into memory if the app has not used any type from the dll.
This causes the framework to fail finding the assembly when it tries to look for it in the list of loaded assemblies.

The fix is to detect this in webassembly and try to load the root component type at that point. Same with the component parameters.

Fixes #52129

## Customer Impact

A Blazor Web app may experience a failure trying to render a root component in WebAssembly, when the component is defined in a Razor Class Library. This will happen if the assembly containing the component hasn't yet been loaded on the client, and the app is trying to render that component.

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

The assembly will already be downloaded by JS and is ready to be loaded.

## Verification

- [x] Manual (required)
- [x] Automated

There was manual verification that the fix addressed the issue and it was automated into an E2E test.

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

There are no packaging changes associated with this fix.

## When servicing release/2.1

- [ ] Make necessary changes in eng/PatchConfig.props

### CODE REVIEW

File: src/Components/Shared/src/ComponentParametersTypeCache.cs [hunk 1]
```diff
@@ -40,10 +40,26 @@ internal sealed class ComponentParametersTypeCache
...
+            // be loaded.
+            if (OperatingSystem.IsBrowser())
+            {
+                try
+                {
+                    assembly = Assembly.Load(key.Assembly);
+                }
+                catch
+                {
+                    // It's fine to ignore the exception, since we'll return null below.
```
Comment (SteveSandersonMS) [inline src/Components/Shared/src/ComponentParametersTypeCache.cs:57]: Might the debugging experience be clearer if we allowed this to be unhandled? Then the developer will know the problem was with assembly loading, as opposed to some other issue with deserializing the parameter markers.

What is the error experience if we do ignore this exception?

**Review by SteveSandersonMS — APPROVED** [review]
In the case where this works, this looks perfect!

My only question is about the debugging experience in the case where there's an assembly loading error. Will leave it with you to decide whether it's already ideal or could be made clearer by letting the exception be thrown.

File: src/Components/Shared/src/RootComponentTypeCache.cs [hunk 2]
```diff
@@ -41,10 +41,26 @@ internal sealed class RootComponentTypeCache
...
+            // be loaded.
+            if (OperatingSystem.IsBrowser())
+            {
+                try
+                {
+                    assembly = Assembly.Load(key.Assembly);
+                }
+                catch
+                {
+                    // It's fine to ignore the exception, since we'll return null below.
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Components/Shared/src/RootComponentTypeCache.cs:58]: https://github.com/dotnet/aspnetcore/pull/52351#discussion_r1404394567

**Comment by javiercn** [comment]
> My only question is about the debugging experience in the case where there's an assembly loading error. Will leave it with you to decide whether it's already ideal or could be made clearer by letting the exception be thrown.

I'm fine leaving it as is (you'll get the same experience you get today), since this is typically a "build up" issue, that happens as you start growing your app. In pretty much every app, you are going to actually reference the assembly types in code explicitly, for example to register services, etc.

So I think this is a point in time issue during dev that just happens when you add the first component and try to render it.

This was/is likely also present in previous webassembly versions, we just didn't make it easy to prerender.

**Comment by SteveSandersonMS** [comment]
> I'm fine leaving it as is (you'll get the same experience you get today), since this is typically a "build up" issue, that happens as you start growing your app. In pretty much every app, you are going to actually reference the assembly types in code explicitly, for example to register services, etc.

To be honest I don't quite get this. Even during this "build up" phase, won't it be easier to understand and fix the `Assembly.Load` failure if we allow it to throw? What's the advantage of intentionally suppressing the exception so that a different error occurs later?

**Comment by javiercn** [comment]
> To be honest I don't quite get this. Even during this "build up" phase, won't it be easier to understand and fix the Assembly.Load failure if we allow it to throw? What's the advantage of intentionally suppressing the exception so that a different error occurs later?

How can this error happen?

**Comment by SteveSandersonMS** [comment]
> How can this error happen?

I don't know, but if it can't, why did you add a try/catch?

**Comment by javiercn** [comment]
> I don't know, but if it can't, why did you add a try/catch?

Because in case it happens, it doesn't change the existing experience, so the fix is more scoped. Thinking hard, the only case where I think this can happen is:
1) The assembly is lazy loaded (there's a separate bug tracking that fix [here](https://github.com/dotnet/aspnetcore/issues/26601)).
2) The assembly is not referenced by the client application at all. (Like you forgot to add the project reference).

In both cases you'll end up with the current error message that tells you that the assembly couldn't be found, which I think it's better than whatever `Assembly.Load` could throw if it fails.

**Pull request merged**